core/outbox_test.go

```diff
@@ -0,0 +1,199 @@
+
+		cid, err := ob.Send(context.Background(), sender, sender, types.NewAttoFILFromFIL(2), types.NewGasPrice(0), types.NewGasUnits(0), "")
+		assert.Errorf(t, err, "for testing")
+		assert.False(t, cid.Defined())
+	})
+
+	t.Run("send message enqueues and calls publish", func(t *testing.T) {
+		w, _ := types.NewMockSignersAndKeyInfo(1)
+		sender := w.Addresses[0]
+		toAddr := address.NewForTestGetter()()
+		queue := core.NewMessageQueue()
+		publisher := &mockPublisher{}
+		provider := &fakeProvider{}
+
+		blk := types.NewBlockForTest(nil, 1)
+		blk.Height = 1000
+		actr, _ := account.NewActor(types.NewZeroAttoFIL())
+		actr.Nonce = 42
+		provider.Set(t, blk, sender, actr)
+
+		ob := core.NewOutbox(w, nullValidator{}, queue, publisher, provider, provider)
+		require.Empty(t, queue.List(sender))
+		require.Nil(t, publisher.message)
+
+		_, err := ob.Send(context.Background(), sender, toAddr, types.NewZeroAttoFIL(), types.NewGasPrice(0), types.NewGasUnits(0), "")
+		require.NoError(t, err)
+		assert.Equal(t, uint64(1000), queue.List(sender)[0].Stamp)
+		assert.NotNil(t, publisher.message)
+		assert.Equal(t, actr.Nonce, publisher.message.Nonce)
+		assert.Equal(t, uint64(1000), publisher.height)
+	})
+
+	t.Run("send message avoids nonce race", func(t *testing.T) {
+		ctx := context.Background()
+		msgCount := 20      // number of messages to send
+		sendConcurrent := 3 // number of of concurrent message sends
+
+		w, _ := types.NewMockSignersAndKeyInfo(1)
+		sender := w.Addresses[0]
+		toAddr := address.NewForTestGetter()()
+		queue := core.NewMessageQueue()
+		publisher := &mockPublisher{}
+		provider := &fakeProvider{}
+
+		blk := types.NewBlockForTest(nil, 1)
+		blk.Height = 1000
+		actr, _ := account.NewActor(types.NewZeroAttoFIL())
+		actr.Nonce = 42
+		provider.Set(t, blk, sender, actr)
+
+		s := core.NewOutbox(w, nullValidator{}, queue, publisher, provider, provider)
+
+		var wg sync.WaitGroup
+		addTwentyMessages := func(batch int) {
+			defer wg.Done()
+			for i := 0; i < msgCount; i++ {
+				_, err := s.Send(ctx, sender, toAddr, types.NewZeroAttoFIL(), types.NewGasPrice(0), types.NewGasUnits(0), fmt.Sprintf("%d-%d", batch, i), []byte{})
+				require.NoError(t, err)
+			}
+		}
+
+		// Add messages concurrently.
+		for i := 0; i < sendConcurrent; i++ {
+			wg.Add(1)
+			go addTwentyMessages(i)
+		}
+		wg.Wait()
+
+		enqueued := queue.List(sender)
+		assert.Equal(t, 60, len(enqueued))
+
+		// Expect the nonces to be distinct and contiguous
+		nonces := map[uint64]bool{}
+		for _, message := range enqueued {
+			assert.Equal(t, uint64(1000), message.Stamp)
+			_, found := nonces[uint64(message.Msg.Nonce)]
+			require.False(t, found)
+			nonces[uint64(message.Msg.Nonce)] = true
+		}
+
+		for i := 0; i < 60; i++ {
+			assert.True(t, nonces[uint64(actr.Nonce)+uint64(i)])
+
+		}
+	})
+
+	t.Run("nonce fails with non-account actor", func(t *testing.T) {
```

Comment:
Should this read "send fails with non-account actor"? If it fails in nextNonce, it's opaque to the test.